api/envoy/config/route/v3/route_components.proto

```diff
@@ -769,10 +769,10 @@ message RouteAction {
     // is used.
     google.protobuf.Duration max_stream_duration = 1;
 
-    // If present, and the request contains a `grpc-timeout` header
+    // If present, and the request contains a `grpc-timeout header
```

Comment:
Looks like you accidentally removed the closing back-tick here.

Reply:
The backtick comes after the link.  (This is how it was done in the old docstrings.)

Reply:
while implementing this, checking if a behavior change was intentional

Previously in Envoy, if this was set and the grpc-timeout header was not present,  Grpc::Common::getGrpcTimeout in FilterUtility::finalTimeout would return 0 (infinite) and the value of grpc_max_timeout would be applied to such gRPC requests.

As I read this, if the grpc-timeout header is not present, grpc_max_timeout will be ignored for that gRPC request and max_stream_duration will be used instead.  Was that the intent?

Reply:
> As I read this, if the grpc-timeout header is not present, grpc_max_timeout will be ignored for that gRPC request and max_stream_duration will be used instead. Was that the intent?

Yes.  The intent here is that the `grpc-timeout` header should affect only the timeout from the start of the RPC (`max_stream_duration`), not the time from the end of the client's half-close (Envoy's normal timeouts, set by the `timeout` field).  The old behavior of modifying the timeout from the client's half-close was broken for gRPC.

Reply:
I think you are not answering the question I asked (unless I'm misreading your answer)?

The behavioral change I'm asking about is not about the time in the point in the life of a request where the grpc-timeout header applies, it is which timeout we want to use if a grpc-timeout header is completely absent.

Reply:
Let me try to clarify.  Envoy has two different, completely independent timeouts:

1. The timeout from the client's half-close, controlled by the `timeout` field.
2. The timeout from the start of the stream, controlled by the `max_stream_duration` field.

I assume that the implementation has a separate timer for each of these, and the timers are started at different points during the stream's lifecycle.  The two timers have nothing to do with each other.  Whichever one fires first causes the request to be terminated.

gRPC's timeout semantics are those of (2), not (1).  We do not have any timer for (1).  However, the existing `max_grpc_timeout` and `grpc_timeout_offset` fields in `RouteAction` modify (1), which makes no sense, because the `grpc-timeout` header is based on (2), not (1).

For example, if a gRPC client sets a timeout of 20 minutes on a bidi streaming RPC, that means that the entire stream should be killed 20 minutes after it starts, even if neither side has half-closed before that.  Envoy's current semantics won't do that, because it won't start the 20-minute timer until the client sends a half-close.

The goals here are:
- Have a way to set a default for (2) on a per-route basis (i.e., overriding `max_stream_duration` in the route).  This effectively allows setting a default timeout for gRPC requests on a per-method basis, which was not really possible before.
- Have a way to dynamically set (2) when the client explicitly specifies one via the `grpc-timeout` header.
- Deprecate the incorrect functionality of using the `grpc-timeout` header to affect (1), since those two things have nothing to do with each other.

Once this change is done, someone who wants to use an infinite gRPC timeout for a long-running stream will need to explicitly set the Envoy `timeout` field to 0 to avoid that timer prematurely breaking the RPC.  There is no default for `max_stream_duration`, so that won't come into play.  And the new `grpc_max_timeout` field won't come into play either, since the request won't have a `grpc-timeout` header (because the client will not set a deadline).

I hope this makes the intent here clearer.  Please let me know if I'm still not making sense.  (We can also chat about it offline if you'd like.)